internal/expr/bindtypes.go

```diff
@@ -75,63 +72,68 @@ func (pe *ParsedExpr) BindTypes(args ...any) (te *TypedExpr, err error) {
 		}
 	}
 
-	var sql bytes.Buffer
-
-	var inCount int
-	var outCount int
-
-	var outputs = make([]typeinfo.Member, 0)
-	var inputs = make([]typeinfo.Member, 0)
-
-	var typeMemberPresent = make(map[typeinfo.Member]bool)
+	// Bind types to each expression.
+	typedExprs := []typedExpression{}
+	outputMemberUsed := map[typeinfo.Member]bool{}
+	for _, expr := range *pe {
+		te, err := expr.bindTypes(ti)
+		if err != nil {
+			return nil, err
+		}
 
-	// Check and expand each query expr.
-	for _, expr := range pe.exprs {
-		switch e := expr.(type) {
-		case *inputExpr:
-			inLoc, err := bindInputTypes(ti, e)
+		if toe, ok := te.(*typedOutputExpr); ok {
+			err = toe.checkUsed(outputMemberUsed)
 			if err != nil {
 				return nil, err
 			}
-			sql.WriteString("@sqlair_" + strconv.Itoa(inCount))
-			inCount++
-			inputs = append(inputs, inLoc)
-		case *outputExpr:
-			outCols, typeMembers, err := bindOutputTypes(ti, e)
-			if err != nil {
-				return nil, err
-			}
-
-			for _, tm := range typeMembers {
-				if ok := typeMemberPresent[tm]; ok {
-					return nil, fmt.Errorf("member %q of type %q appears more than once in output expressions", tm.MemberName(), tm.OuterType().Name())
-				}
-				typeMemberPresent[tm] = true
-			}
-
-			for i, c := range outCols {
-				sql.WriteString(c.String())
-				sql.WriteString(" AS ")
-				sql.WriteString(markerName(outCount))
-				if i != len(outCols)-1 {
-					sql.WriteString(", ")
-				}
-				outCount++
-			}
-			outputs = append(outputs, typeMembers...)
-		case *bypass:
-			sql.WriteString(e.chunk)
-		default:
-			return nil, fmt.Errorf("internal error: unknown query expr type %T", expr)
 		}
+
+		typedExprs = append(typedExprs, te)
 	}
 
-	return &TypedExpr{inputs: inputs, outputs: outputs, sql: sql.String()}, nil
+	typedExpr := TypedExprs(typedExprs)
+	return &typedExpr, nil
 }
 
-// bindInputTypes binds the input expression to a type and returns the
-// typeMember represented by the expression.
-func bindInputTypes(ti typeNameToInfo, e *inputExpr) (tm typeinfo.Member, err error) {
+// expression represents a parsed node of the SQLair query's AST.
+type expression interface {
+	// String returns a text representation for debugging and testing purposes.
+	String() string
+
+	// bindTypes generates a typed expression from the query type information.
+	bindTypes(typeNameToInfo) (typedExpression, error)
+}
+
+// bypass represents part of the expression that we want to pass to the backend
```

Comment:
I will remove the word `part`. Maybe something like:
```go
// bypass represents an expression that we want to pass to the backend database verbatim

// bypass is a node of the AST that we want to pass to the backend database verbatim
```
I prefer the second one.

Reply:
Yea, this sentence could do with rewording. I've changed it to
```
// bypass is an expression representing a chunk of SQL that we want to pass to
// the backend database verbatim.
```
I want to talk about it in terms of the types we have rather than as a node in the AST, even though it is that. We have the types above the we can uses as names for it to make it clear both how we use it and what it is at the same time.